### ISSUE TITLE
False positive for "Warning: Redundant void"

### ISSUE THREAD

**Issue by jezen** [ndmitchell/hlint#836]
I have this line in a function in my project

```
forM_ chats $ void . runMessagingT env . uncurry (prepareRoom registeredUsers)
```

The hlint tool suggested the following change

```
backend/src/Network/Chat/Xmpp.hs:145:3: Warning: Redundant void
Found:
  forM_ chats $
    void . runMessagingT env . uncurry (prepareRoom registeredUsers)
Perhaps:
  forM_ chats
    (runMessagingT env . uncurry (prepareRoom registeredUsers))
```

But this resulted in the following type error

```
src/Network/Chat/Xmpp.hs:145:17: error:
    • Couldn't match type ‘((), Stream)’ with ‘()’
      Expected type: Element [(Entity Chat, [Entity User])] -> m ()
        Actual type: (Entity Chat, [Entity User]) -> m ((), Stream)
    • In the second argument of ‘($)’, namely
        ‘runMessagingT env . uncurry (prepareRoom registeredUsers)’
      In a stmt of a 'do' block:
        forM_ chats
          $ runMessagingT env . uncurry (prepareRoom registeredUsers)
      In the expression:
        do registeredUsers <- getRegisteredUsersForHost (VHost xmeServer)
                                >>= logEjabberdAPIErr []
           when (xmppLogin `notElem` registeredUsers)
             $ do let ...
                  res <- liftIO $ registerNewUser euser vhost
                  ....
           forM_ chats
             $ runMessagingT env . uncurry (prepareRoom registeredUsers)
           eiResult <- runMessagingT env
                         $ connect xmppLogin xmppPass
                             >>=
                               \case
                                 Right me -> forM_ chats (curry enterRoom user me) $> Right me
                                 Left err -> pure $ Left err
           ....
    |
145 |   forM_ chats $ runMessagingT env . uncurry (prepareRoom registeredUsers)
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Language extensions enabled in this file are the following

```
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE FlexibleContexts  #-}
{-# LANGUAGE RecordWildCards   #-}
{-# LANGUAGE ConstraintKinds   #-}
{-# LANGUAGE TemplateHaskell   #-}
{-# LANGUAGE TupleSections     #-}
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE RankNTypes        #-}
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE GADTs             #-}
```

**Comment by josephcsible** [comment]
Here's the relevant hint: https://github.com/ndmitchell/hlint/blob/f404c85ef5e14bff6e0216072e04ea040821fbbd/data/hlint.yaml#L444

Is there any chance you can post a minimal, complete program that demonstrates the problem, or at least the definition of `MessagingT`? I'm guessing that it's an exotic type (either higher-rank, existential, or a GADT).

**Comment by ndmitchell** [comment]
Are you using a different `forM_`? Looking through Hoogle  Data.MonoTraversable.Unprefixed would do it, but the vast majority of `forM_` things should work.

**Comment by jezen** [comment]
@ndmitchell We are indeed not using the `forM_` from `base`.

```
λ :i forM_
forM_ ::
  (MonoFoldable mono, Applicative m) =>
  mono -> (Element mono -> m ()) -> m ()
        -- Defined in ‘Data.MonoTraversable.Unprefixed’
```

@josephcsible not sure if it's still helpful given the above, but our `MessagingT` is like this:

```
type MessagingT m a = ReaderT XmppEnvironment (XmppMonad m) a
```

**Comment by jezen** [comment]
FYI, this is possibly related to https://github.com/ndmitchell/hlint/issues/541.

**Comment by ndmitchell** [comment]
Indeed, the same one. In both cases I'd just recommend disabling the warning for now - a proper fix will take some time.

**Comment by jezen** [comment]
It's not a problem for me or my team, I'm just trying to be a _good citizen_ and file issues when they come up. If there is some plan of action elsewhere (perhaps in haskell/ide?), then it might be best to close this and #541. What do you think? How would you best like to organise this?

**Comment by ndmitchell** [comment]
Thanks, I very much appreciate it! I'll link to the plan of action from the other ticket.
